### ISSUE TITLE
Issue 006: coarse infer に 2D GPT を接続する（single-tile / no channel tiling）

### ISSUE THREAD

**Issue by ozw4** [ozw4/seisai#355]
# Issue 006: coarse infer に 2D GPT を接続する（single-tile / no channel tiling）

## 背景

Issue 001-003 で path_tracing core と segmented DP2D が実装され、Issue 004 で config が入り、Issue 005 で `.coarse.npz` diagnostics を保存できるようになった。この issue では、まず channel tiling が実質 1 tile のケースに限定して、coarse inference へ GPT を接続する。

channel tiling が複数 tile になる場合は、tile ごとの GPT ではなく、tile probability map を先に merge してから GPT する必要がある。これは Issue 007 で対応する。この issue では、複数 tile のときに GPT enabled なら明確な error を出して、安全でない tile-local GPT を避ける。

## 目的

`run_coarse_infer()` で以下を実現する。

```text
model logits
  -> log_softmax score
  -> segment-aware 2D GPT
  -> GPT 後 anchor pick
  -> raw time projection
  -> restore to full traces
  -> .coarse.npz required keys に GPT 後 pick を保存
  -> optional diagnostics に argmax / GPT comparison を保存
```

## 非目的

この issue では以下は行わない。

- 複数 channel tile の score merge
- QC overlay 更新
- 3D GPT
- physics/fine の変更
- raw waveform / pilot GPT

## 変更対象

```text
packages/seisai-engine/src/seisai_engine/pipelines/fbpick/coarse/infer.py
packages/seisai-engine/tests/test_fbpick_coarse.py
packages/seisai-engine/tests/test_fbpick_coarse_time_axis.py  # 必要なら
```

### 1. helper 分離

`coarse/infer.py` に helper を追加するか、読みやすければ `coarse/gpt_infer.py` を新設する。

推奨 helper:

```python
def _extract_anchor_predictions_argmax(
    logits_hw: np.ndarray,
) -> tuple[np.ndarray, np.ndarray]:
    ...
```

既存 `_extract_anchor_predictions()` は argmax 用として残すか rename する。

GPT 用:

```python
def _trace_anchor_predictions_with_gpt(
    logits_hw: np.ndarray,
    *,
    trace_valid: np.ndarray,
    segment_id: np.ndarray,
    cfg: GlobalPathTracingCfg,
) -> tuple[np.ndarray, np.ndarray, PathTracingResult]:
    ...
```

挙動:

- `logits_hw` を `log_softmax_time(logits_hw, time_axis=-1)` に変換する。
- `prob_hw = prob_from_log_score(score_hw)`。
- `ScoreVolume(score=score_hw, prob=prob_hw, valid_mask=trace_valid, axes=("trace", "time"), time_axis=1)` を作る。
- `PathTopology(kind="segment_1d", component_id=segment_id, valid_mask=trace_valid)` を作る。
- `trace_global_path()` を呼ぶ。
- 戻り値:
  - `anchor_pick_j`: `result.path_j`
  - `anchor_pmax`: `result.path_prob`
  - `result`

### 2. run_coarse_infer flow

現状:

```python
anchor_pick_j, anchor_pmax = _extract_anchor_predictions(logits_bhw[batch_idx])
```

を以下に置き換える。

```python
argmax_pick_j, argmax_pmax = _extract_anchor_predictions_argmax(logits_hw)

if typed.infer.global_path_tracing.enabled:
    anchor_pick_j, anchor_pmax, gpt_result = _trace_anchor_predictions_with_gpt(...)
else:
    anchor_pick_j, anchor_pmax = argmax_pick_j, argmax_pmax
    gpt_result = None
```

### 3. invalid row handling

- `trace_valid=False` row は `anchor_pick_j_validated=-1`。
- GPT result の invalid row も `-1`。
- valid row で `anchor_pmax` が `nan` にならないことを確認する。
- `project_coarse_indices_to_raw_time(... ignore_index=-1)` は既存通り使う。

### 4. single tile restriction

GPT enabled のとき、dataset に複数 tile item がある場合は、この issue では明確に fail する。

実装候補:

```python
if typed.infer.global_path_tracing.enabled:
    if any(int(item.tile_count) > 1 for item in dataset.items):
        raise ValueError(
            "global_path_tracing with channel tiling requires tiled score merge; "
            "disable infer.channel_tiling or implement Issue 007"
        )
```

Issue 007 でこの restriction を削除する。

### 5. full trace diagnostics

single-tile では、argmax と GPT の両方を restore できる。

```python
restored_pick_i_argmax, restored_pmax_argmax = restore_anchor_predictions_to_full_traces(... argmax ...)
restored_pick_i_gpt, restored_pmax_gpt = restore_anchor_predictions_to_full_traces(... GPT ...)
```

final:

```python
coarse_pick_i = restored_pick_i_gpt if GPT enabled else restored_pick_i_argmax
coarse_pmax = restored_pmax_gpt if GPT enabled else restored_pmax_argmax
```

saved `extra_arrays` when GPT enabled and diagnostics true:

```text
coarse_pick_i_argmax
coarse_pmax_argmax
coarse_pick_i_gpt
coarse_pmax_gpt
coarse_gpt_adjust_samples = coarse_pick_i_gpt - coarse_pick_i_argmax
coarse_gpt_component_id   # restored/interpolated or nearest segment id per trace
coarse_gpt_valid_mask     # bool, final GPT valid
coarse_center_valid_mask  # bool, downstream center validity
coarse_gpt_enabled
coarse_gpt_algorithm
coarse_gpt_score_type
coarse_gpt_max_jump_coarse_samples
coarse_gpt_min_component_anchors
```

`coarse_gpt_component_id` の restore は MVP では以下でよい。

- segment metadata から raw trace position ごとの segment id vector を作る。
- dtype `int32`。

`coarse_center_valid_mask` は GPT enabled の single-tile では基本的に all True になる。将来欠損が出る場合に備えて vector として保存する。

### 6. backward compatibility

`global_path_tracing.enabled == False` の場合:

- 既存 output と同じ required keys。
- optional GPT keys は保存しない。
- 既存 tests が通る。

## テスト要件

追加 / 更新:

```python
def test_run_coarse_infer_gpt_disabled_matches_argmax_behavior() -> None: ...
def test_run_coarse_infer_gpt_enabled_saves_gpt_required_keys_as_main() -> None: ...
def test_run_coarse_infer_gpt_enabled_saves_argmax_diagnostics() -> None: ...
def test_run_coarse_infer_gpt_enabled_uses_logsoftmax_not_probability_sum() -> None: ...
def test_run_coarse_infer_gpt_enabled_rejects_multiple_channel_tiles_until_merge() -> None: ...
```

実データ不要の fake dataset / fake model test が難しい場合は、既存 test style に合わせて helper 単位でテストしてよい。

最低限、`_trace_anchor_predictions_with_gpt()` に対して synthetic logits を作り、row-wise argmax では飛ぶが GPT では smooth path になることを確認する。

## 受け入れ条件

- `infer.global_path_tracing.enabled=true` で single-tile coarse infer が GPT 後 pick を `.coarse.npz` main key に保存する。
- argmax diagnostics が optional keys に保存される。
- `enabled=false` の既存挙動は維持される。
- 複数 tile では tile-local GPT をせず、明確な error を出す。

## 推奨確認コマンド

```bash
pytest packages/seisai-engine/tests/test_fbpick_path_tracing_core.py \
       packages/seisai-engine/tests/test_fbpick_path_tracing_dp2d.py \
       packages/seisai-engine/tests/test_fbpick_path_tracing_segmented.py \
       packages/seisai-engine/tests/test_fbpick_coarse.py -q
```
